contracts/c-sphincs-all-in-one-lock/ckb-sphincsplus-root-lock.c

```diff
@@ -0,0 +1,406 @@
+
+  blake2b_state s;
+  init_blake2b_state(&s, PERSONAL_SCRIPT);
+
+  BytesOptType lock = witness_args->t->lock(witness_args);
+  CHECK2(!lock.t->is_none(&lock), ERROR_SPHINCSPLUS_WITNESS);
+
+  mol2_cursor_t lock_bytes = lock.t->unwrap(&lock);
+
+  uint8_t multisig_headers[4];
+  CHECK(mol2_read_and_advance(&lock_bytes, multisig_headers, 4));
+
+  mol2_cursor_t signatures = lock_bytes;
+
+  CHECK2(multisig_headers[0] == MULTISIG_RESERVED_FIELD_VALUE,
+         ERROR_SPHINCSPLUS_ENCODING);
+  uint8_t require_first_n = multisig_headers[1];
+  uint8_t threshold = multisig_headers[2];
+  uint8_t pubkeys = multisig_headers[3];
+  CHECK2(pubkeys > 0, ERROR_SPHINCSPLUS_ENCODING);
+  CHECK2(threshold <= pubkeys, ERROR_SPHINCSPLUS_ENCODING);
+  CHECK2(threshold > 0, ERROR_SPHINCSPLUS_ENCODING);
+  CHECK2(require_first_n <= threshold, ERROR_SPHINCSPLUS_ENCODING);
+
+  blake2b_update(&s, multisig_headers, 4);
+
+  int last_param_id_state = PARAM_ID_STATE_UNSET;
+  int last_param_id = 0;
+
+  for (uint8_t i = 0; i < pubkeys; i++) {
+    uint8_t id;
+    CHECK(mol2_read_and_advance(&lock_bytes, &id, 1));
+    uint8_t param_id = id & MULTISIG_PARAMS_ID_MASK;
+    blake2b_update(&s, &param_id, 1);
+
+    const CkbSphincsParams *params = NULL;
+    CHECK(fetch_params(param_id, &params));
+
+    uint8_t pubkey[params->pk_bytes];
+    CHECK(mol2_read_and_advance(&lock_bytes, pubkey, params->pk_bytes));
+    blake2b_update(&s, pubkey, params->pk_bytes);
+
+    if ((id & MULTISIG_SIG_MASK) != 0) {
+      mol2_advance(&lock_bytes, params->sign_bytes);
+
+      CHECK2(threshold > 0, ERROR_SPHINCSPLUS_WITNESS);
+      threshold--;
+    } else {
+      CHECK2(i >= require_first_n, ERROR_SPHINCSPLUS_WITNESS);
+    }
+
+    /*
+     * This is actually a process to dedup param IDs from all
+     * public keys, to see if all public keys use a single param
+     * ID.
+     */
+    switch (last_param_id_state) {
+      case PARAM_ID_STATE_UNSET: {
+        last_param_id = param_id;
+        last_param_id_state = PARAM_ID_STATE_SET;
+      } break;
+      case PARAM_ID_STATE_SET: {
+        if (last_param_id != param_id) {
+          last_param_id = 0;
+          last_param_id_state = PARAM_ID_STATE_MULTIPLE;
+        }
+      } break;
+      case PARAM_ID_STATE_MULTIPLE: {
+        /* No-op here*/
+      } break;
+      default: {
+        return ERROR_SPHINCSPLUS_UNEXPECTED;
+      } break;
+    }
+  }
+
+  CHECK2(threshold == 0, ERROR_SPHINCSPLUS_WITNESS);
+  CHECK2(lock_bytes.size == 0, ERROR_SPHINCSPLUS_WITNESS);
+  blake2b_final(&s, actual_multisig_hash, BLAKE2B_BLOCK_SIZE);
+  *out_signatures = signatures;
+  *out_params_id = last_param_id;
+
+exit:
+  return err;
+}
+
+int main(int argc, char *argv[]) {
+  int err = CKB_SUCCESS;
+
+  /* Extract current script's data first, so we can reclaim buffer set aside
+   for loading script to be used later */
+  uint8_t code_hash[BLAKE2B_BLOCK_SIZE];
+  uint8_t hash_type;
+  uint8_t expected_multisig_hash[BLAKE2B_BLOCK_SIZE];
```

Comment:
The script size can affect CKB capacity locked. Can we have 20 bytes args instead of 32?

Reply:
This is a deliberate design choice here: we've been saying that using blake160 saves us 12 bytes for so long, but we have never confirmed the necessity to save those 12 bytes. So for a security oriented lock like this SPHINCS+ one, I would favor full 32 bytes blake2b hash.